src/lsptoolshost/onAutoInsertFeature.ts

```diff
@@ -0,0 +1,137 @@
+    }
+
+    public getState(): FeatureState {
+        const selectors = this.getDocumentSelectors();
+
+        let count = 0;
+        for (const selector of selectors) {
+            count++;
+            for (const document of Workspace.textDocuments) {
+                if (Languages.match(selector, document) > 0) {
+                    return { kind: 'document', id: this.registrationType.method, registrations: true, matches: true };
+                }
+            }
+        }
+        const registrations = count > 0;
+        return { kind: 'document', id: this.registrationType.method, registrations, matches: false };
+    }
+
+    public fillClientCapabilities(capabilities: ClientCapabilities): void {
+        const textDocumentCapabilities: any = ensure(capabilities, 'textDocument')!;
+        if (textDocumentCapabilities['_vs_onAutoInsert'] === undefined) {
+            textDocumentCapabilities['_vs_onAutoInsert'] = {} as any;
+        }
+        const onAutoInsertCapability = textDocumentCapabilities['_vs_onAutoInsert'];
+        onAutoInsertCapability.dynamicRegistration = true;
+    }
+
+    public initialize(_capabilities: ServerCapabilities, documentSelector: DocumentSelector): void {
+        const capabilities: any = _capabilities;
+        const options = this.getRegistrationOptions(documentSelector, capabilities._vs_onAutoInsertProvider);
+        if (!options) {
+            return;
+        }
+        this.register({
+            id: UUID.generateUuid(),
+            registerOptions: options,
+        });
+    }
+
+    public getOptions(textDocument: TextDocument): RoslynProtocol.OnAutoInsertOptions | undefined {
+        for (const registration of this._registrations.values()) {
+            const selector = registration.registerOptions.documentSelector;
+            if (
+                selector !== null &&
+                Languages.match(this._client.protocol2CodeConverter.asDocumentSelector(selector), textDocument) > 0
+            ) {
+                return registration.registerOptions;
+            }
+        }
+        return undefined;
+    }
+
+    private *getDocumentSelectors(): IterableIterator<VDocumentSelector> {
+        for (const registration of this._registrations.values()) {
+            const selector = registration.registerOptions.documentSelector;
+            if (selector === null) {
+                continue;
+            }
+            yield this._client.protocol2CodeConverter.asDocumentSelector(selector);
+        }
+    }
+
+    private getRegistrationOptions(
+        documentSelector: DocumentSelector | undefined,
+        capability: undefined | boolean | RoslynProtocol.OnAutoInsertOptions
+    ): (RoslynProtocol.OnAutoInsertRegistrationOptions & { documentSelector: DocumentSelector }) | undefined {
+        if (!documentSelector || !capability) {
+            return undefined;
+        }
+        return (
+            Is.boolean(capability) && capability === true
```

Comment:
Doubtful but TextDocumentLanguageFeature does it like this so I kept it.

Reply:
I think that is because there are a lot of capabilities that are either booleans or an object - https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#serverCapabilities

Since we control this end to end (and its never a boolean), I believe we should remove the boolean check